docs/explanation/testing.md

```diff
@@ -1,67 +1,115 @@
 
-{ref}`testenv]
+- System operations such as writing files.
+- Cloud operations such as launching virtual machines.
+- Workload operations, using Pebble in the case of a Kubernetes charm.
+- Juju operations such as sharing data with related charms.
 
-basepython = python3.10
--->
+Unit tests focus on mapping these inputs to expected outputs. For example, a unit test could verify a system call, the contents of a file, or the contents of a relation databag.
 
+> See also: {ref}`write-scenario-tests-for-a-charm`.
 
-## Unit testing
+### Coverage
 
-> See also: {ref}`write-legacy-unit-tests-for-a-charm`, {ref}`write-scenario-tests-for-a-charm`
+Unit testing a charm should cover:
 
-A **unit test** is a test that targets an individual unit of code (function, method, class, etc.) independently. In the context of a charm, it refers to testing charm code against mock Juju APIs and mocked-out workloads as a way to validate isolated behaviour without external interactions.
+- How relation data is modified as a result of an event.
+- What pebble services are running as a result of an event.
+- Which configuration files are written and their contents, as a result of an event.
 
-Unit tests are intended to be isolating and fast to complete. These are the tests you would run every time before committing code changes.
+### Tools
 
-**Coverage.** Unit testing a charm should cover:
+- [`ops.testing`](ops_testing), the framework for state-transition testing in Ops
+- [`pytest`](https://pytest.org/) and/or [`unittest`](https://docs.python.org/3/library/unittest.html)
+- [`tox`](https://tox.wiki/en/latest/index.html) for automating and standardizing tests
 
-- how relation data is modified as a result of an event
-- what pebble services are running as a result of an event
-- which configuration files are written and their contents, as a result of an event
+The `ops.testing` framework provides `State`, which mocks inputs and outputs. The framework also provides `Context` and `Container`, which offer mock filesystems. Tests involve:
 
-**Tools.** Unit testing a charm can be done using:
+1. Setting up the charm, metadata, context, output mocks, and Juju state.
+2. Simulating events using `Context.run`. For example, `config_changed`, `relation_changed`, `storage_attached`, `pebble_ready`, and so on.
+3. Retrieving and asserting the output.
 
-- [`pytest`](https://pytest.org/) and/or [`unittest`](https://docs.python.org/3/library/unittest.html) and
-- [state transition testing](ops_testing), using the `ops` unit testing framework
+`Context` and `State` are instantiated before the charm. This enables you to prepare the state of config, relations, and storage before simulating an event.
 
-**Examples.**
+### Examples
 
 - [https://github.com/canonical/prometheus-k8s-operator/blob/main/tests/unit/test_charm.py](https://github.com/canonical/prometheus-k8s-operator/blob/main/tests/unit/test_charm.py)
```

Comment:
Let's update this example. It uses legacy `ops.testing.Harness` rather than scenario / state-transition testing. Can we link to one with scenario-based tests? Also, the link should be labelled, like `[Prometheus K8s charm unit tests](...)`